File: Arrays & Strings/edit_distance.py
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        Example 1:

    Input: word1 = "horse", word2 = "ros"
    Output: 3
        """
        matrix = [[0 for __ in range(len(word1)+1)] for _ in range(len(word2)+1)]
        for i in range(len(word1)+1):matrix[0][i] = i
        for i in range(len(word2)+1):matrix[i][0] = i
        
        for i in range(1,len(word2)+1):
            for j in range(1,len(word1)+1):
                matrix[i][j] = 1 + min(matrix[i][j-1],matrix[i-1][j-1],matrix[i-1][j]) if word2[i-1] != word1[j-1] else matrix[i-1][j-1]
        return matrix[-1][-1]
        

File: java/DP/edit_distance.java
class Solution {
    public int minDistance(String word1, String word2) {
        int[][] matrix = new int[word1.length()+1][word2.length()+1];
        
        for (int i=0;i<word1.length()+1;i++) matrix[i][0] = i;
        for (int i=0;i<word2.length()+1;i++) matrix[0][i] = i;
        
        for (int i=1; i<word1.length()+1; i++){
            for (int j=1; j<word2.length()+1; j++){
                if (word1.charAt(i-1)!=word2.charAt(j-1))
                    matrix[i][j] = 1 + Math.min(Math.min(matrix[i-1][j],matrix[i-1][j-1]),matrix[i][j-1]);
                else
                    matrix[i][j] = matrix[i-1][j-1];
            }
        }
        
        return matrix[word1.length()][word2.length()];
    }
}
